src/transformStateWithClones.js

```diff
@@ -6,8 +6,37 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const obj = { ...state };
+
+  for (const object of actions) {
+    const { extraData, keysToRemove } = object;
+
+    if (object.type === 'addProperties') {
+      for (const key in extraData) {
+        obj[key] = extraData[key];
+      }
+      stateHistory.push({ ...obj });
+    }
+
+    if (object.type === 'removeProperties') {
+      for (const key in keysToRemove) {
+        delete obj[keysToRemove[key]];
+      }
```

Comment:
The `keysToRemove` is expected to be an array, not an object. Therefore, you should iterate over it using `for (const key of keysToRemove)` instead of `for (const key in keysToRemove)`. This will correctly iterate over the array elements, which are the keys to be removed.